ee/server/service/hosts.go

```diff
@@ -19,3 +21,105 @@ func (svc *Service) HostByIdentifier(ctx context.Context, identifier string, opt
 	opts.IncludePolicies = true
 	return svc.Service.HostByIdentifier(ctx, identifier, opts)
 }
+
+func (svc *Service) OSVersions(ctx context.Context, teamID *uint, platform *string, name *string, version *string, opts fleet.ListOptions) (*fleet.OSVersions, int, *fleet.PaginationMetadata, error) {
+	var count int
+	if err := svc.authz.Authorize(ctx, &fleet.Host{TeamID: teamID}, fleet.ActionList); err != nil {
+		return nil, count, nil, err
+	}
+
+	if name != nil && version == nil {
+		return nil, count, nil, &fleet.BadRequestError{Message: "Cannot specify os_name without os_version"}
+	}
+
+	if name == nil && version != nil {
+		return nil, count, nil, &fleet.BadRequestError{Message: "Cannot specify os_version without os_name"}
+	}
+
+	osVersions, err := svc.ds.OSVersions(ctx, teamID, platform, name, version)
+	if err != nil && fleet.IsNotFound(err) {
+		// differentiate case where team was added after UpdateOSVersions last ran
+		if teamID != nil && *teamID > 0 {
+			// most of the time, team should exist so checking here saves unnecessary db calls
+			_, err := svc.ds.Team(ctx, *teamID)
+			if err != nil {
+				return nil, count, nil, err
+			}
+		}
+		// if team exists but stats have not yet been gathered, return empty JSON array
+		osVersions = &fleet.OSVersions{}
+	} else if err != nil {
+		return nil, count, nil, err
+	}
+
+	for i, os := range osVersions.OSVersions {
+
+		// populate OSVersion.GeneratedCPEs
+		if os.Platform == "darwin" {
+			osVersions.OSVersions[i].GeneratedCPEs = []string{
+				fmt.Sprintf("cpe:2.3:o:apple:macos:%s:*:*:*:*:*:*:*", os.Version),
+				fmt.Sprintf("cpe:2.3:o:apple:mac_os_x:%s:*:*:*:*:*:*:*", os.Version),
+			}
+		}
+
+		// populate OSVersion.Vulnerabilities
+		vulns, err := svc.ds.ListVulnsByOS(ctx, os.ID, true)
+		if err != nil {
+			return nil, count, nil, err
+		}
+
+		osVersions.OSVersions[i].Vulnerabilities = make(fleet.Vulnerabilities, 0) // avoid null in JSON
+		for _, vuln := range vulns {
+			switch os.Platform {
+			case "darwin":
+				vuln.DetailsLink = fmt.Sprintf("https://nvd.nist.gov/vuln/detail/%s", vuln.CVE)
+			case "windows":
+				vuln.DetailsLink = fmt.Sprintf("https://msrc.microsoft.com/update-guide/en-US/vulnerability/%s", vuln.CVE)
+			}
+			osVersions.OSVersions[i].Vulnerabilities = append(osVersions.OSVersions[i].Vulnerabilities, vuln)
+		}
+	}
+
+	if opts.OrderKey == "hosts_count" && opts.OrderDirection == fleet.OrderAscending {
```

Comment:
Is `hosts_count` the only allowed order key? If so, please add validation if needed.